osu.Game/Tests/Visual/OnlinePlay/OnlinePlayTestScene.cs

```diff
@@ -67,7 +68,24 @@ protected sealed override IReadOnlyDependencyContainer CreateChildDependencies(I
             // To get around this, the BeatmapManager is looked up from the dependencies provided to the children of the test scene instead.
             var beatmapManager = dependencies.Get<BeatmapManager>();
 
-            ((DummyAPIAccess)API).HandleRequest = request => handler.HandleRequest(request, API.LocalUser.Value, beatmapManager);
+            ((DummyAPIAccess)API).HandleRequest = request =>
+            {
+                TaskCompletionSource<bool> tcs = new TaskCompletionSource<bool>();
+
+                // Because some of the handlers use realm, we need to ensure the game is still alive when firing.
+                // If we don't, a stray `PerformAsync` could hit an `ObjectDisposedException` if running too late.
+                Scheduler.Add(() =>
+                {
+                    bool result = handler.HandleRequest(request, API.LocalUser.Value, beatmapManager);
+                    tcs.SetResult(result);
+                }, false);
+
+#pragma warning disable RS0030
+                // We can't GetResultSafely() here (will fail with "Can't use GetResultSafely from inside an async operation."), but Wait is safe enough due to
+                // the task being a TaskCompletionSource.
+                return tcs.Task.Result;
```

Comment:
I was actually surprised that this doesn't deadlock single-threaded, but then realised that it doesn't because the schedule is using `Scheduler.Add(..., false)`, i.e. it doesn't force the schedule if the thread is already correct. Maybe worth an inline comment?

Reply:
Can do.